### PR TITLE
Added e2e test to change pvc template

### PR DESCRIPTION
**Pull request by d-kuro**
refs: https://github.com/cybozu-go/moco/issues/265

depends: https://github.com/cybozu-go/moco/pull/403

Implement e2e test for pvc template changes and pvc resize functionality.

### CODE REVIEW

File: controllers/pvc.go [hunk 1]
```diff
@@ -58,17 +58,22 @@ func (r *MySQLClusterReconciler) reconcilePVC(ctx context.Context, req ctrl.Requ
...
-	if err := r.resizePVCs(ctx, cluster, &sts, resizeTarget); err != nil {
+	resized, err := r.resizePVCs(ctx, cluster, &sts, resizeTarget)
+	if err != nil {
 		metrics.VolumeResizedErrorTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
 		return err
 	}
 
+	if len(resized) == 0 {
+		return nil
+	}
```
Comment (d-kuro) [inline controllers/pvc.go:69]: There was a problem that when StatefulSet reconcile failed for some reason, the metric was counted up even though no resizing was done, so we decided to count the PVC that was actually resized.

File: controllers/mysqlcluster_controller.go [hunk 2]
```diff
@@ -953,9 +953,9 @@ func (r *MySQLClusterReconciler) reconcileV1StatefulSet(ctx context.Context, req
 					return false, err
 				}
 
-				return false, fmt.Errorf("re-creation failed the StatefulSet %s/%s has not been deleted", cluster.Namespace, cluster.PrefixedName())
+				return false, nil
 			}); err != nil {
-				return err
+				return fmt.Errorf("re-creation failed the StatefulSet %s/%s has not been deleted: %w", cluster.Namespace, cluster.PrefixedName(), err)
```
Comment (d-kuro) [inline controllers/mysqlcluster_controller.go:958]: BugFix: It was not retried

File: controllers/pvc.go [hunk 3]
```diff
@@ -197,6 +208,11 @@ func (r *MySQLClusterReconciler) isVolumeExpansionSupported(ctx context.Context,
 // isUpdatingStatefulSet returns whether the StatefulSet is being updated or not.
 // refs: https://github.com/kubernetes/kubectl/blob/v0.24.2/pkg/polymorphichelpers/rollout_status.go#L119-L152
 func (*MySQLClusterReconciler) isUpdatingStatefulSet(sts *appsv1.StatefulSet) bool {
+	// Waiting for StatefulSet to be deleting.
+	if sts.DeletionTimestamp != nil {
+		return true
+	}
```
Comment (d-kuro) [inline controllers/pvc.go:214]: Fixed to not try to resize PVC if StatefulSet is being deleted.

File: controllers/pvc.go [hunk 4]
```diff
@@ -124,17 +131,21 @@ func (r *MySQLClusterReconciler) resizePVCs(ctx context.Context, cluster *mocov1
...
+			return resizedPVC, fmt.Errorf("failed to resize pvc %q, want size: %s, deployed size: %s: %w", pvc.Name, newSize.String(), pvc.Spec.Resources.Requests.Storage().String(), ErrReduceVolumeSize)
 		case i == -1: // current volume size is smaller than new size
 			pvc.Spec.Resources.Requests[corev1.ResourceStorage] = *newSize
 
 			if err := r.Client.Update(ctx, &pvc); err != nil {
-				return fmt.Errorf("failed to update PVC: %w", err)
+				return resizedPVC, fmt.Errorf("failed to update PVC: %w", err)
 			}
+
+			log.Info("PVC resized", "namespace", pvc.Namespace, "pvcName", pvc.Name)
```
[the remark this answers is not in the document]
Comment (d-kuro) [reply controllers/pvc.go]: Thanks, I removed it.
https://github.com/cybozu-go/moco/pull/417/commits/4f8e2d814fcf1aa61221fd6ef7a81fc569ed2fa2

**Review by masa213f — APPROVED** [review]
LGTM!

**Pull request merged**